File: ui/ts/bench-tab.ts
import { invoke } from "@tauri-apps/api/core";

export interface BenchResponse {
  throughput_mbps: number;
  latency_us: number;
  bytes_generated: number;
  duration_secs: number;
}

export function initBenchTab() {
  const sizeSlider = document.getElementById('bench-size-slider') as HTMLInputElement;
  const sizeLabel = document.getElementById('bench-size-label') as HTMLElement;
  const runBtn = document.getElementById('bench-run-btn') as HTMLButtonElement;
  const resultsSection = document.getElementById('bench-results-section') as HTMLElement;

  if (!sizeSlider || !runBtn || !resultsSection) return;

  const updateSliderLabel = () => {
    // Logarithmic: 10,000 to 10,000,000 => 10^4 to 10^7
    const val = parseFloat(sizeSlider.value);
    const bytes = Math.round(Math.pow(10, val));
    sizeLabel.textContent = bytes.toLocaleString() + " bytes";
    return bytes;
  };

  sizeSlider.addEventListener('input', updateSliderLabel);
  updateSliderLabel(); // initial

  let isLoading = false;

  const runBenchmark = async () => {
    if (isLoading) return;
    const bytes = updateSliderLabel();

    isLoading = true;
    runBtn.disabled = true;
    runBtn.textContent = "Benchmarking...";
    resultsSection.classList.add('hidden');

    try {
      const response = await invoke<BenchResponse>('run_benchmark', { bytes });

      document.getElementById('bench-throughput-val')!.textContent = `${response.throughput_mbps.toFixed(1)} MB/s`;
      document.getElementById('bench-latency-val')!.textContent = `${response.latency_us.toFixed(2)} µs/byte`;
      document.getElementById('bench-bytes-val')!.textContent = response.bytes_generated.toLocaleString();
      document.getElementById('bench-duration-val')!.textContent = `${response.duration_secs.toFixed(2)} s`;

      resultsSection.classList.remove('hidden');

    } catch (e) {
      console.error(e);
      alert(`Error running benchmark: ${e}`);
    } finally {
      isLoading = false;
      runBtn.disabled = false;
      runBtn.textContent = "⚡ Run Benchmark";
    }
  };

  runBtn.addEventListener('click', runBenchmark);
}

File: src-tauri/src/commands/bench_tab.rs
use serde::{Serialize, Deserialize};
use entropy_forge::bench::PerformanceBench;
use entropy_forge::entropy::SystemEntropy;

#[derive(Serialize, Deserialize)]
pub struct BenchResponse {
    pub throughput_mbps: f64,
    pub latency_us: f64,
    pub bytes_generated: usize,
    pub duration_secs: f64,
}

#[tauri::command]
pub fn run_benchmark(bytes: usize) -> BenchResponse {
    let mut entropy = SystemEntropy::new();
    let result = PerformanceBench::benchmark(&mut entropy, bytes);

    BenchResponse {
        throughput_mbps: result.throughput_mbps,
        latency_us: result.latency_us,
        bytes_generated: result.bytes_generated,
        duration_secs: result.duration.as_secs_f64(),
    }
}
